evmrpc/state.go

```diff
@@ -0,0 +1,85 @@
+package evmrpc
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/sei-protocol/sei-chain/x/evm/keeper"
+	rpcclient "github.com/tendermint/tendermint/rpc/client"
+)
+
+type StateAPI struct {
+	tmClient    rpcclient.Client
+	keeper      *keeper.Keeper
+	ctxProvider func() sdk.Context
+}
+
+func NewStateAPI(tmClient rpcclient.Client, k *keeper.Keeper, ctxProvider func() sdk.Context) *StateAPI {
+	return &StateAPI{tmClient: tmClient, keeper: k, ctxProvider: ctxProvider}
+}
+
+func (a *StateAPI) GetBalance(ctx context.Context, address common.Address, blockNr rpc.BlockNumber) (*uint64, error) {
+	block, err := getBlockNumber(ctx, a.tmClient, blockNr)
+	if err != nil {
+		return nil, err
+	}
+	if block != nil {
+		return nil, errors.New("block number not safe, finalized, latest, or pending")
+	}
+
+	balance := a.keeper.GetBalance(a.ctxProvider(), address)
```

Comment:
`GetBalance` gets balance stored in `evm` module - only accounts that haven't established the association between their evm address and their sei address will have their balance stored in `evm` module. Otherwise balance is stored in the bank module. So we need to first call `a.keeper.GetSeiAddress` to check if there is an association, and use `a.keeper.BankKeeper().GetBalance(ctx, seiAddr, a.keeper.GetBaseDenom(ctx))` to get the bank balance of the account if so. If there is no association then we use `a.keeper.GetBalance`